src/realm/object-store/sync/sync_session.cpp

```diff
@@ -232,73 +234,124 @@ SyncSession::handle_refresh(const std::shared_ptr<SyncSession>& session)
-    std::lock_guard<std::mutex> lk(m_state_mutex);
+    util::CheckedLockGuard lk(m_state_mutex);
     m_sync_manager = nullptr;
 }
 
-std::string SyncSession::get_recovery_file_path()
-{
-    return util::reserve_unique_file_name(m_sync_manager->recovery_directory_path(m_config.recovery_directory),
-                                          util::create_timestamped_template("recovered_realm"));
-}
-
 void SyncSession::update_error_and_mark_file_for_deletion(SyncError& error, ShouldBackup should_backup)
 {
+    util::CheckedLockGuard config_lock(m_config_mutex);
     // Add a SyncFileActionMetadata marking the Realm as needing to be deleted.
     std::string recovery_path;
     auto original_path = path();
     error.user_info[SyncError::c_original_file_path_key] = original_path;
     if (should_backup == ShouldBackup::yes) {
-        recovery_path = get_recovery_file_path();
+        recovery_path =
+            util::reserve_unique_file_name(m_sync_manager->recovery_directory_path(m_config.recovery_directory),
+                                           util::create_timestamped_template("recovered_realm"));
         error.user_info[SyncError::c_recovery_file_path_key] = recovery_path;
     }
     using Action = SyncFileActionMetadata::Action;
     auto action = should_backup == ShouldBackup::yes ? Action::BackUpThenDeleteRealm : Action::DeleteRealm;
-    m_sync_manager->perform_metadata_update([this, action, original_path = std::move(original_path),
-                                             recovery_path = std::move(recovery_path)](const auto& manager) {
-        std::string partition_value = m_config.partition_value;
-        manager.make_file_action_metadata(original_path, partition_value, m_config.user->identity(), action,
-                                          recovery_path);
+    m_sync_manager->perform_metadata_update(
+        [action, original_path = std::move(original_path), recovery_path = std::move(recovery_path),
+         partition_value = m_config.partition_value, identity = m_config.user->identity()](const auto& manager) {
+            manager.make_file_action_metadata(original_path, partition_value, identity, action, recovery_path);
+        });
+}
+
+
+void SyncSession::download_fresh_realm()
+{
+    DBOptions options;
+    options.encryption_key = m_db->get_encryption_key();
+    options.allow_file_format_upgrade = false;
+    options.enable_async_writes = false;
+
+    std::shared_ptr<DB> db;
+    auto fresh_path = ClientResetOperation::get_fresh_path_for(m_db->get_path());
+    try {
+        // We want to attempt to use a pre-existing file to reduce the chance of
+        // downloading the first part of the file only to then delete it over
+        // and over, but if we fail to open it then we should just start over.
+        try {
+            db = DB::create(sync::make_client_replication(), fresh_path, options);
+        }
+        catch (...) {
+            util::File::try_remove(fresh_path);
+        }
```

Comment:
This bit should be the one functional change from the previous version of the code which used RealmCoordinator/AsyncOpenTask.